com.unity.render-pipelines.high-definition/Runtime/Water/WaterVertexTessellation.hlsl

```diff
@@ -0,0 +1,225 @@
+    edgeTessFactors *= distFactor * distFactor;
+
+    edgeTessFactors *= inputTessellationFactors * _GlobalTessellationFactorMultiplier;
+
+    // TessFactor below 1.0 have no effect. At 0 it kill the triangle, so clamp it to 1.0
+    edgeTessFactors = max(edgeTessFactors, float3(1.0, 1.0, 1.0));
+
+    return CalcTriTessFactorsFromEdgeTessFactors(edgeTessFactors);
+}
+
+struct TessellationFactors
+{
+    float edge[3] : SV_TessFactor;
+    float inside : SV_InsideTessFactor;
+};
+
+TessellationFactors HullConstant(InputPatch<PackedVaryingsToDS, 3> input)
+{
+    VaryingsToDS varying0 = UnpackVaryingsToDS(input[0]);
+    VaryingsToDS varying1 = UnpackVaryingsToDS(input[1]);
+    VaryingsToDS varying2 = UnpackVaryingsToDS(input[2]);
+
+    float3 p0 = varying0.vmesh.positionRWS;
+    float3 p1 = varying1.vmesh.positionRWS;
+    float3 p2 = varying2.vmesh.positionRWS;
+
+    float3 n0 = varying0.vmesh.normalWS;
+    float3 n1 = varying1.vmesh.normalWS;
+    float3 n2 = varying2.vmesh.normalWS;
+
+    // x - 1->2 edge
+    // y - 2->0 edge
+    // z - 0->1 edge
+    // w - inside tessellation factor (calculate as mean of three in GetTessellationFactors())
+    float3 inputTessellationFactors;
+    // TessellatinFactor is evaluate in vertex shader
+    inputTessellationFactors.x = 0.5 * (varying1.vmesh.tessellationFactor + varying2.vmesh.tessellationFactor);
+    inputTessellationFactors.y = 0.5 * (varying2.vmesh.tessellationFactor + varying0.vmesh.tessellationFactor);
+    inputTessellationFactors.z = 0.5 * (varying0.vmesh.tessellationFactor + varying1.vmesh.tessellationFactor);
+
+    float4 tf = GetTessellationFactors(p0, p1, p2, n0, n1, n2, inputTessellationFactors);
+
+    TessellationFactors output;
+    output.edge[0] = min(tf.x, MAX_TESSELLATION_FACTORS);
+    output.edge[1] = min(tf.y, MAX_TESSELLATION_FACTORS);
+    output.edge[2] = min(tf.z, MAX_TESSELLATION_FACTORS);
+    output.inside  = min(tf.w, MAX_TESSELLATION_FACTORS);
+
+    return output;
+}
+
+// ref: http://reedbeta.com/blog/tess-quick-ref/
+[maxtessfactor(MAX_TESSELLATION_FACTORS)]
+[domain("tri")]
+[partitioning("fractional_odd")]
+[outputtopology("triangle_cw")]
+[patchconstantfunc("HullConstant")]
+[outputcontrolpoints(3)]
+PackedVaryingsToDS Hull(InputPatch<PackedVaryingsToDS, 3> input, uint id : SV_OutputControlPointID)
+{
+    // Pass-through
+    return input[id];
+}
+
+[domain("tri")]
+PackedVaryingsToPS Domain(TessellationFactors tessFactors, const OutputPatch<PackedVaryingsToDS, 3> input, float3 baryCoords : SV_DomainLocation)
+{
+    VaryingsToDS varying0 = UnpackVaryingsToDS(input[0]);
+    VaryingsToDS varying1 = UnpackVaryingsToDS(input[1]);
+    VaryingsToDS varying2 = UnpackVaryingsToDS(input[2]);
+
+    VaryingsToDS varying = InterpolateWithBaryCoordsToDS(varying0, varying1, varying2, baryCoords);
+
+    return VertTesselation(varying);
```

Comment:
note for myself. Lot of duplication, can we avoid it?

Reply:
I don't think we can without making it a mess